File: src/brs/db/sql/SqlAliasStore.java
package brs.db.sql;

import brs.Alias;
import brs.Signum;
import brs.db.SignumKey;
import brs.db.VersionedEntityTable;
import brs.db.store.AliasStore;
import brs.db.store.DerivedTableManager;
import brs.util.Convert;

import org.jooq.Condition;
import org.jooq.DSLContext;
import org.jooq.Record;
import org.jooq.Record1;
import org.jooq.Result;
import org.jooq.SortField;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Locale;

import static brs.schema.Tables.ALIAS;
import static brs.schema.Tables.ALIAS_OFFER;

public class SqlAliasStore implements AliasStore {

  private static final DbKey.LongKeyFactory<Alias.Offer> offerDbKeyFactory = new DbKey.LongKeyFactory<Alias.Offer>(ALIAS_OFFER.ID) {
      @Override
      public SignumKey newKey(Alias.Offer offer) {
        return offer.dbKey;
      }
    };

  public SqlAliasStore(DerivedTableManager derivedTableManager) {
    offerTable = new VersionedEntitySqlTable<Alias.Offer>("alias_offer", ALIAS_OFFER, offerDbKeyFactory, derivedTableManager) {
      @Override
      protected Alias.Offer load(DSLContext ctx, Record record) {
        return new SqlOffer(record);
      }

      @Override
      protected void save(DSLContext ctx, Alias.Offer offer) {
        saveOffer(offer);
      }
    };

    aliasTable = new VersionedEntitySqlTable<Alias>("alias", brs.schema.Tables.ALIAS, aliasDbKeyFactory, derivedTableManager) {
      @Override
      protected Alias load(DSLContext ctx, Record record) {
        return new SqlAlias(record);
      }

      @Override
      protected void save(DSLContext ctx, Alias alias) {
        saveAlias(ctx, alias);
      }

      @Override
      protected List<SortField<?>> defaultSort() {
        List<SortField<?>> sort = new ArrayList<>();
        sort.add(tableClass.field("alias_name_lower", String.class).asc());
        return sort;
      }
    };
  }

  @Override
  public SignumKey.LongKeyFactory<Alias.Offer> getOfferDbKeyFactory() {
    return offerDbKeyFactory;
  }

  private static final SignumKey.LongKeyFactory<Alias> aliasDbKeyFactory = new DbKey.LongKeyFactory<Alias>(ALIAS.ID) {

      @Override
      public SignumKey newKey(Alias alias) {
        return alias.dbKey;
      }
    };

  @Override
  public SignumKey.LongKeyFactory<Alias> getAliasDbKeyFactory() {
    return aliasDbKeyFactory;
  }

  @Override
  public VersionedEntityTable<Alias> getAliasTable() {
    return aliasTable;
  }

  private class SqlOffer extends Alias.Offer {
    private SqlOffer(Record record) {
      super(record.get(ALIAS_OFFER.ID), record.get(ALIAS_OFFER.PRICE), Convert.nullToZero(record.get(ALIAS_OFFER.BUYER_ID)), offerDbKeyFactory.newKey(record.get(ALIAS_OFFER.ID)));
    }
  }

  private void saveOffer(Alias.Offer offer) {
    Db.useDSLContext(ctx -> {
      ctx.insertInto(ALIAS_OFFER, ALIAS_OFFER.ID, ALIAS_OFFER.PRICE, ALIAS_OFFER.BUYER_ID, ALIAS_OFFER.HEIGHT)
              .values(offer.getId(), offer.getPriceNqt(), (offer.getBuyerId() == 0 ? null : offer.getBuyerId()), Signum.getBlockchain().getHeight())
              .execute();
    });
  }

  private final VersionedEntityTable<Alias.Offer> offerTable;

  @Override
  public VersionedEntityTable<Alias.Offer> getOfferTable() {
    return offerTable;
  }

  private class SqlAlias extends Alias {
    private SqlAlias(Record record) {
      super(
            record.get(ALIAS.ID),
            record.get(ALIAS.ACCOUNT_ID),
            record.get(ALIAS.ALIAS_NAME),
            record.get(ALIAS.TLD),
            record.get(ALIAS.ALIAS_URI),
            record.get(ALIAS.TIMESTAMP),
            aliasDbKeyFactory.newKey(record.get(ALIAS.ID))
            );
    }
  }

  private void saveAlias(DSLContext ctx, Alias alias) {
    ctx.insertInto(ALIAS).
      set(ALIAS.ID, alias.getId()).
      set(ALIAS.ACCOUNT_ID, alias.getAccountId()).
      set(ALIAS.ALIAS_NAME, alias.getAliasName()).
      set(ALIAS.TLD, alias.getTld()).
      set(ALIAS.ALIAS_NAME_LOWER, alias.getAliasName().toLowerCase(Locale.ENGLISH)).
      set(ALIAS.ALIAS_URI, alias.getAliasUri()).
      set(ALIAS.TIMESTAMP, alias.getTimestamp()).
      set(ALIAS.HEIGHT, Signum.getBlockchain().getHeight()).execute();
  }

  private final VersionedEntityTable<Alias> aliasTable;

  @Override
  public Collection<Alias> getAliasesByOwner(long accountId, String name, Long tld, int from, int to) {
    Condition condition = ALIAS.TLD.isNotNull();
    if(tld != null) {
      condition = ALIAS.TLD.eq(tld);
    }
    if(accountId != 0L) {
      condition = condition.and(ALIAS.ACCOUNT_ID.eq(accountId));
    }
    if(name != null) {
      condition = condition.and(ALIAS.ALIAS_NAME_LOWER.like(name.toLowerCase()));
    }
    return aliasTable.getManyBy(condition, from, to);
  }

  @Override
  public Collection<Alias> getTLDs(int from, int to) {
    return aliasTable.getManyBy(brs.schema.Tables.ALIAS.TLD.isNull(), from, to);
  }

  @Override
  public Collection<Alias.Offer> getAliasOffers(long account, long buyer, int from, int to) {
    Condition conditions = ALIAS_OFFER.LATEST.eq(true);
    if(account != 0L) {
      Result<Record1<Long>> myAliases = Db.useDSLContext(ctx -> {
      return ctx.select(ALIAS.ID).from(ALIAS).where(ALIAS.ACCOUNT_ID.eq(account)).fetch();
      });
      conditions = conditions.and(ALIAS_OFFER.ID.in(myAliases));
    }
    if(buyer != 0L) {
      conditions = conditions.and(ALIAS_OFFER.BUYER_ID.eq(buyer));
    }
    return offerTable.getManyBy(conditions, from, to);
  }
  
  @Override
  public Alias getTLD(String tldName) {
    return aliasTable.getBy(brs.schema.Tables.ALIAS.ALIAS_NAME_LOWER.eq(tldName.toLowerCase(Locale.ENGLISH)).and(ALIAS.TLD.isNull()));
  }

  @Override
  public Alias getTLD(long tldId) {
    return aliasTable.getBy(brs.schema.Tables.ALIAS.ID.eq(tldId).and(ALIAS.TLD.isNull()));
  }

  @Override
  public Alias getAlias(String aliasName, long tld) {
    return aliasTable.getBy(brs.schema.Tables.ALIAS.ALIAS_NAME_LOWER.eq(aliasName.toLowerCase(Locale.ENGLISH)).and(ALIAS.TLD.eq(tld)));
  }

}

File: src/main/java/brs/db/sql/SqlAliasStore.kt
package brs.db.sql

import brs.Alias
import brs.DependencyProvider
import brs.db.BurstKey
import brs.db.VersionedEntityTable
import brs.db.store.AliasStore
import brs.schema.Tables.ALIAS
import brs.schema.Tables.ALIAS_OFFER
import brs.schema.tables.records.AliasOfferRecord
import brs.schema.tables.records.AliasRecord
import brs.util.Convert
import org.jooq.DSLContext
import org.jooq.Record
import org.jooq.SortField
import java.util.*

class SqlAliasStore(private val dp: DependencyProvider) : AliasStore {
    override val offerTable: VersionedEntityTable<Alias.Offer>
    override val aliasTable: VersionedEntityTable<Alias>
    override val aliasDbKeyFactory = AliasDbKeyFactory
    override val offerDbKeyFactory = OfferDbKeyFactory

    init {
        offerTable = object : VersionedEntitySqlTable<Alias.Offer>("alias_offer", ALIAS_OFFER, offerDbKeyFactory, dp) {
            override fun load(ctx: DSLContext, record: Record): Alias.Offer {
                return SqlOffer(record)
            }

            override fun save(ctx: DSLContext, offer: Alias.Offer) {
                saveOffer(offer)
            }
        }

        aliasTable = object : VersionedEntitySqlTable<Alias>("alias", brs.schema.Tables.ALIAS, aliasDbKeyFactory, dp) {
            override fun load(ctx: DSLContext, record: Record): Alias {
                return SqlAlias(record)
            }

            override fun save(ctx: DSLContext, alias: Alias) {
                saveAlias(ctx, alias)
            }

            override fun defaultSort(): List<SortField<*>> {
                val sort = mutableListOf<SortField<*>>()
                sort.add(tableClass.field("alias_name_lower", String::class.java).asc())
                return sort
            }
        }
    }

    private inner class SqlOffer internal constructor(record: Record) : Alias.Offer(record.get(ALIAS_OFFER.ID), record.get(ALIAS_OFFER.PRICE), Convert.nullToZero(record.get(ALIAS_OFFER.BUYER_ID)), offerDbKeyFactory.newKey(record.get(ALIAS_OFFER.ID)))

    private fun saveOffer(offer: Alias.Offer) {
        Db.useDSLContext { ctx ->
            ctx.insertInto<AliasOfferRecord, Long, Long, Long, Int>(ALIAS_OFFER, ALIAS_OFFER.ID, ALIAS_OFFER.PRICE, ALIAS_OFFER.BUYER_ID, ALIAS_OFFER.HEIGHT)
                    .values(offer.id, offer.priceNQT, if (offer.buyerId == 0L) null else offer.buyerId, dp.blockchain.height)
                    .execute()
        }
    }

    private inner class SqlAlias internal constructor(record: Record) : Alias(record.get(ALIAS.ID), record.get(ALIAS.ACCOUNT_ID), record.get(ALIAS.ALIAS_NAME), record.get(ALIAS.ALIAS_URI), record.get(ALIAS.TIMESTAMP), aliasDbKeyFactory.newKey(record.get(ALIAS.ID)))

    private fun saveAlias(ctx: DSLContext, alias: Alias) {
        ctx.insertInto<AliasRecord>(ALIAS).set(ALIAS.ID, alias.id).set(ALIAS.ACCOUNT_ID, alias.accountId).set(ALIAS.ALIAS_NAME, alias.aliasName).set(ALIAS.ALIAS_NAME_LOWER, alias.aliasName.toLowerCase(Locale.ENGLISH)).set(ALIAS.ALIAS_URI, alias.aliasURI).set(ALIAS.TIMESTAMP, alias.timestamp).set(ALIAS.HEIGHT, dp.blockchain.height).execute()
    }

    override fun getAliasesByOwner(accountId: Long, from: Int, to: Int): Collection<Alias> {
        return aliasTable.getManyBy(brs.schema.Tables.ALIAS.ACCOUNT_ID.eq(accountId), from, to)
    }

    override fun getAlias(aliasName: String): Alias? {
        return aliasTable.getBy(brs.schema.Tables.ALIAS.ALIAS_NAME_LOWER.eq(aliasName.toLowerCase(Locale.ENGLISH)))
    }

    companion object {
        object OfferDbKeyFactory : DbKey.LongKeyFactory<Alias.Offer>(ALIAS_OFFER.ID) {
            override fun newKey(offer: Alias.Offer): BurstKey {
                return offer.dbKey
            }
        }

        object AliasDbKeyFactory : DbKey.LongKeyFactory<Alias>(ALIAS.ID) {
            override fun newKey(alias: Alias): BurstKey {
                return alias.dbKey
            }
        }
    }
}
